### PR TITLE
fix: Remove old data provider listener on component attach

### PR DESCRIPTION
**Pull request by mshabarov**
## Description

This change is a fix for ComboBox and related to the fix made on the flow side to get rid of memory leaks caused by old listeners being hanging in the list - https://github.com/vaadin/flow/pull/11471.

Null-value checking of `dataProviderListener` is not needed in attach handler because:
1. Calling to a `remove()` is more robust and this is a normal way of handling listeners registrations in flow. Additionally, if, for some reason, a new data provider will be set up to the component, it's more correct to remove the old listener sitting in the old data provider.
2. Not calling of `remove()` method gives an inconsistency with the flow, which now (with the above fix) removes the listener on attach event.

The inconsistency described in 2. causes the ComboBox's and DataCommunicator data provider listeners order violation.
This causes a side-effects in client-side filtering, it's seen from the failed tests.
I didn't dig deeply into the details what a the side effect in deep, but it's clear that it is somehow overrides/cleans filter/items/size variables in data communicator -> break it's internal state.

How it worked before both fixes:
1. DataCommunicator adds a data provider listener upon its `setDataProvider` method.
2. ComboBox adds a data provider listener upon its `setDataProvider` method.
3. DataCommunicator adds one more data provider listener upon attach event.
4. ComboBox adds one more data provider listener upon attach event.

So finally DC listener comes before CB listener and this work fine.

How it worked after the fix in Flow:
1. DataCommunicator adds a data provider listener upon its `setDataProvider` method.
2. ComboBox adds a data provider listener upon its `setDataProvider` method.
3. DataCommunicator REMOVES the old data provider listener upon attach event and adds a new one.
4. ComboBox adds one more data provider listener upon attach event.

So now the CB listener comes first (because the first listener in the queue has been removed), and this violates the order of listener execution. And this causes a side-effects and test failures.

How it works after the fix in Flow and ComboBox:
1. DataCommunicator adds a data provider listener upon its `setDataProvider` method.
2. ComboBox adds a data provider listener upon its `setDataProvider` method.
3. DataCommunicator REMOVES the old data provider listener upon attach event and adds a new one.
4. ComboBox REMOVES the old data provider listener upon attach event and adds a new one.

Thus, there are only two listeners in the end and they are in the right order -> works fine.

No tests added, because this change is covered by tests in `FilteringIT`.

Related-to https://github.com/vaadin/flow-components/issues/1914

## Type of change

- [x] Bugfix
- [ ] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs-beta/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [ ] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

File: vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBox.java [hunk 1]
```diff
@@ -942,7 +942,7 @@ private void executeRegistration(PropertyChangeEvent event) {
     protected void onAttach(AttachEvent attachEvent) {
         super.onAttach(attachEvent);
         DataProvider<T, ?> dataProvider = getDataProvider();
-        if (dataProvider != null && dataProviderListener == null) {
+        if (dataProvider != null) {
```
Comment (pleku) [inline vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBox.java:945]: Unit test please ?

**Review by pleku — CHANGES_REQUESTED** [review]

File: vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBox.java (hunk 1 above)
Comment (mshabarov) [reply vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBox.java:945]: Initially, I was thinking that this change is covered by `FilteringIT`. But this filtering test is a separate thing, and it's quite hard to make connections between filtering and DP listeners. So, it's a good idea to add a specific test covering this particular case with DP listeners.

Comment (mshabarov) [reply vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBox.java:945]: The `Grid` has the same code (not leading to a test failures though), so I'll fix it as well.

Comment (mshabarov) [reply vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBox.java:945]: And some other component which use data provider...:)

Comment (mshabarov) [reply vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/java/com/vaadin/flow/component/combobox/ComboBox.java:945]: Done

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java [hunk 2]
```diff
@@ -101,4 +109,66 @@ public void setAllRowsVisibleProperty_isHeightByRowsAndIsAllRowsVisibleWork() {
...
+                Collections.emptyList()) {
+            @Override
+            public Registration addDataProviderListener(
+                    DataProviderListener<String> listener) {
+                listenersCount.incrementAndGet();
+                Registration registration = super.addDataProviderListener(
+                        listener);
+                // the first listener is added by Grid in 'setDataProvider'
+                // and it is the one we want to be removed.
+                return listenersCount.get() == 1
```
Comment (pleku) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java:130]: I'm trying to wrap my head around why is this needed for the test, but cannot, so please elaborate ? I understand collecting the number of listeners present, but don't understand why this changes the logic of the method directly. Seems odd.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java [hunk 3]
```diff
@@ -101,4 +109,66 @@ public void setAllRowsVisibleProperty_isHeightByRowsAndIsAllRowsVisibleWork() {
...
+                "Expected data provider listener added by Grid not to be "
+                        + "removed before Grid being attached",
+                oldListenerRemoved.get());
+
+        // given
+        listenersCount.set(0);
+
+        // when
+        DataCommunicatorTest.MockUI mockUI = new DataCommunicatorTest.MockUI();
+        mockUI.add(grid);
```
Comment (pleku) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java:155]: As the detach is also removing the listener(s) and it is a critical part of keeping things working, is there another unit test that verifies that detach does it or could you possibly also include verifying that detach cleans up the listeners, so that we can guarantee that the next attach will work ?

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java [hunk 4]
```diff
@@ -101,4 +109,66 @@ public void setAllRowsVisibleProperty_isHeightByRowsAndIsAllRowsVisibleWork() {
...
+                // and it is the one we want to be removed.
+                return listenersCount.get() == 1
+                        ? Registration.combine(registration,
+                                () -> oldListenerRemoved.set(true))
+                        : registration;
+            }
+        };
+
+        // when
+        grid.setDataProvider(dataProvider);
```
Comment (pleku) [inline vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java:138]: The logic feels off to me that the listeners are added
a) when the component is attached - OK
b) when the data provider is added - I would expect that here it is only added if the component was attached to begin with

File: vaadin-radio-button-flow-parent/vaadin-radio-button-flow/src/test/java/com/vaadin/flow/component/radiobutton/RadioButtonGroupTest.java [hunk 5]
```diff
@@ -417,4 +424,63 @@ public void addNullOption_setValue() {
...
+        // when
+        DataCommunicatorTest.MockUI mockUI = new DataCommunicatorTest.MockUI();
+        mockUI.add(rbg);
+        fakeClientCommunication(mockUI);
+
+        // then
+        Assert.assertEquals(
+                "Expected exactly one data provider listener added by "
+                        + "RBG's attach handle method",
+                2, listenersCount.get());
```
Comment (pleku) [inline vaadin-radio-button-flow-parent/vaadin-radio-button-flow/src/test/java/com/vaadin/flow/component/radiobutton/RadioButtonGroupTest.java:474]: Again, it verifies for against two but there should be only one ?

**Review by pleku — CHANGES_REQUESTED** [review]
There is a lot of repetition in the unit tests, which probably could be refactored into something reusable so that maintenance could be easier ?

File: vaadin-list-box-flow-parent/vaadin-list-box-flow/src/test/java/com/vaadin/flow/component/listbox/test/ListBoxUnitTest.java [hunk 6]
```diff
@@ -163,6 +171,65 @@ public void setIdentifierProviderOnId_setItemWithNullId_shouldThrowException() {
...
+        // when
+        DataCommunicatorTest.MockUI mockUI = new DataCommunicatorTest.MockUI();
+        mockUI.add(listBox);
+        fakeClientCommunication(mockUI);
+
+        // then
+        Assert.assertEquals(
+                "Expected exactly one data provider listener added by "
+                        + "ListBox's attach handle method",
+                2, listenersCount.get());
```
[the remark this answers is not in the document]
Comment (mshabarov) [reply vaadin-list-box-flow-parent/vaadin-list-box-flow/src/test/java/com/vaadin/flow/component/listbox/test/ListBoxUnitTest.java:220]: Expected count = 2 here is the accumulated value. I will reset the counter before component attach in order to improve the readability.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java (hunk 2 above)
Comment (mshabarov) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java:130]: The goal of doing that is no only check that the given listener has been removed. Since there are several listeners added, including ones added by DataCommunicator, I use the known order of this listener. In case of Grid, the listener added in `setDataProvider` method by Grid is the first one in the list and it is supposed to be removed upon  `onAttach` method.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java (hunk 3 above)
Comment (mshabarov) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java:155]: Yes, good suggestion, I'll check and add if needed.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java (hunk 4 above)
Comment (mshabarov) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java:138]: Basically, I agree that in most cases it's a non sense to call `refreshAll` or `refreshItem` on detached component, because the updated won't be seen anyway. But there are a couple of concerns to do that:
1. Some 'exotic' cases when `refresh` methods are called even before component attach and some logic relies on that.
2. Should we distinguish between 'initial' attach and further re-attach events? I mean, for the initial attach we don't need to add the listener, but should we add the listener upon `setDataProvider` between further reattachments? The logic goes more complex and this is a door for regressions.

File: vaadin-radio-button-flow-parent/vaadin-radio-button-flow/src/test/java/com/vaadin/flow/component/radiobutton/RadioButtonGroupTest.java (hunk 5 above)
Comment (mshabarov) [reply vaadin-radio-button-flow-parent/vaadin-radio-button-flow/src/test/java/com/vaadin/flow/component/radiobutton/RadioButtonGroupTest.java:474]: `listenersCount` will be re-set so as to validate agains expected '1' value, not '2'.

File: vaadin-list-box-flow-parent/vaadin-list-box-flow/src/main/java/com/vaadin/flow/component/listbox/ListBoxBase.java [hunk 7]
```diff
@@ -102,6 +102,7 @@ public void setDataProvider(DataProvider<ITEM, ?> dataProvider) {
         this.dataProvider.set(Objects.requireNonNull(dataProvider));
         DataViewUtils.removeComponentFilterAndSortComparator(this);
         clear();
+        rebuild();
```
Comment (pleku) [inline vaadin-list-box-flow-parent/vaadin-list-box-flow/src/main/java/com/vaadin/flow/component/listbox/ListBoxBase.java:105]: Did you just fix another regression or what is this change about ? 😄 

Wondering if it should be highlighted as an additional fix. Was there a test failure that occurred that caught it ? IT? Or unit test ?

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java (hunk 4 above)
Comment (pleku) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java]: I think it should work in simply following heurestic:
When setting a data provider
a) if component is attached, initialize everything (most things only take effect on before-client-response phase)
b) if component is not attached, wait for it to be attached before doing everything
On detach things are cleaned up as much as possible, but it may wary (currently) between components whether the data is reused or not (probably not). 
On setting a new data provider, a) and b) rules should always apply.

So I don't at the moment see the reason for the more complex logic between initial and reattach (when it comes to data provider things, browser related things are different).

File: vaadin-list-box-flow-parent/vaadin-list-box-flow/src/main/java/com/vaadin/flow/component/listbox/ListBoxBase.java (hunk 7 above)
Comment (mshabarov) [reply vaadin-list-box-flow-parent/vaadin-list-box-flow/src/main/java/com/vaadin/flow/component/listbox/ListBoxBase.java:105]: Good question!
This was another regression, which was revealed when I deleted registration `null` check in `ListBox`.
It fails `ListBoxIT.componentsBetweenItems` tests and the reason was that `rebuild()` method was in the wrong place and it was called mistakenly in `setupDataProviderListener`.
I placed it next to `clear()` method, as it done in the data provider listener.

I can place it into additional fix, but then I'll need to revert all changes in ListBox in this PR.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java (hunk 3 above)
Comment (mshabarov) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java]: Added

File: vaadin-list-box-flow-parent/vaadin-list-box-flow/src/test/java/com/vaadin/flow/component/listbox/test/ListBoxUnitTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mshabarov) [reply vaadin-list-box-flow-parent/vaadin-list-box-flow/src/test/java/com/vaadin/flow/component/listbox/test/ListBoxUnitTest.java]: Reworked.

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java (hunk 4 above)
Comment (mshabarov) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java]: Thanks for clarifications! I'll try to implement this logic.

File: vaadin-list-box-flow-parent/vaadin-list-box-flow/src/main/java/com/vaadin/flow/component/listbox/ListBoxBase.java (hunk 7 above)
Comment (pleku) [reply vaadin-list-box-flow-parent/vaadin-list-box-flow/src/main/java/com/vaadin/flow/component/listbox/ListBoxBase.java:105]: It is fine I think (like this), but please just highlight this in the commit description when merging.

**Review by pleku — APPROVED** [review]

File: vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java (hunk 4 above)
Comment (mshabarov) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java]: @pleku you approved this PR, so is it fine if I implement this suggestion in the separate PR?

Comment (pleku) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java]: Yes I think it should be a separate PR as it is additional and between refactor/bugfix. And another issue should be created as this does handle the original one.

Comment (pleku) [reply vaadin-grid-flow-parent/vaadin-grid-flow/src/test/java/com/vaadin/flow/component/grid/GridTest.java]: And the point about calling `refresh` or `refreshAll` doesn't make sense in my opinion if the component is not attached - since there should not be anything cached (or is there and then why?) then the _refresh_ is a NOOP as all the items will be anyway updated/refetched once the component is attached again.

In case some component retains the fetched items over detach+reattach, that sounds like premature optimization to me.
